src/Protocol/ProtocolRecognizer.cpp

```diff
@@ -1225,22 +222,132 @@ void cProtocolRecognizer::HandlePacketStatusRequest(void)
 
 	AString Response = JsonUtils::WriteFastString(ResponseValue);
 
-	cPacketizer Pkt(*this, pktStatusResponse);
-	Pkt.WriteString(Response);
+	VERIFY(a_Out.WriteVarInt32(GetPacketID(cProtocol::ePacketType::pktStatusResponse)));
+	VERIFY(a_Out.WriteVarUTF8String(Response));
 }
 
 
 
 
 
-void cProtocolRecognizer::HandlePacketStatusPing()
+/* Ping handler for unrecognised versions. */
+static void HandlePacketStatusPing(cClientHandle & a_Client, cByteBuffer & a_Buffer, cByteBuffer & a_Out)
 {
 	Int64 Timestamp;
-	if (!m_Buffer.ReadBEInt64(Timestamp))
+	if (!a_Buffer.ReadBEInt64(Timestamp))
 	{
 		return;
 	}
 
-	cPacketizer Pkt(*this, pktPingResponse);
-	Pkt.WriteBEInt64(Timestamp);
+	VERIFY(a_Out.WriteVarInt32(GetPacketID(cProtocol::ePacketType::pktPingResponse)));
+	VERIFY(a_Out.WriteBEInt64(Timestamp));
+}
+
+
+
+
+
+std::unique_ptr<cProtocol> TryRecogniseProtocol(cClientHandle & a_Client, cByteBuffer & a_SeenData, std::string_view & a_Data)
+{
+	if (!a_SeenData.Write(a_Data.data(), a_Data.size()))
+	{
+		// We read more than the handshake packet here, oh well.
+
+		a_Client.Kick("Too much data was sent");
+		return {};
+	}
+
+	try
+	{
+		auto Protocol = TryRecognizeProtocol(a_Client, a_SeenData);
+		if (Protocol != nullptr)
+		{
+			// The protocol has just been recognized, advance data start
+			// to after the handshake and leave the rest to the protocol:
+			a_Data = a_Data.substr(a_SeenData.GetUsedSpace() - a_SeenData.GetReadableSpace());
+
+			// We read more than we can handle, purge the rest:
+#ifdef NDEBUG
+			a_SeenData.SkipRead(a_SeenData.GetReadableSpace());
+#else
+			ASSERT(a_SeenData.SkipRead(a_SeenData.GetReadableSpace()));
```

Comment:
I don't like the duplication here, how about this instead:
```cpp
[[maybe_unused]] bool Success = a_SeenData.SkipRead(a_SeenData.GetReadableSpace());
ASSERT(Success);
```

Reply:
everyday we stray farther from C